### DISCUSSION TITLE
Sorting of output?

### DISCUSSION THREAD

**Discussion by richmintz1** [todotxt/todo.txt-cli#331] [Q&A] [answered]
I don't understand what mechanism is doing the sorting of output when I "list" items using todo.sh. My specific goal is to be able to sort by due date rather than priority when I need to, but I'd also like to understand how sorting happens in general.

I'm competent to understand a moderately unixy explanation and to make minor modifications to todo.sh.

If this is documented somewhere, just point me there, but I can't find it.

Thanks in advance!

**Comment by inkarkat** [comment, marked as answer]
`TODOTXT_SORT_COMMAND` is mentioned in the default [todo.cfg](https://github.com/todotxt/todo.txt-cli/blob/355166da671404a06fc86a20f593ca9bda68042b/todo.cfg#L81-L85). This is a Bash _command-line_ (a command or a pipeline of commands that typically includes the `sort` command) thats gets the (somewhat preprocessed) tasks via standard input, reorders the lines, and then hands them back via standard output to the colorization stage inside `todo.sh`.

Due dates are typically set via a `t:<date>` or `due:<date>` marker. As these can occur anywhere in the task text, sorting based on the date is a bit difficult. The usual approach is

1. extracting the due date and duplicating it to the beginning of each line (using a default date for tasks that don't have a due date; that default value determines whether non-due tasks sort before or after due tasks)
1. sorting numerically on the first column (optionally considering other columns as well to maintain the default order when multiple tasks have the same due date)
1. removing the added first column again (so that the colorization stage can do the proper highlighting)

So, it will look something like this:

```
export TODOTXT_SORT_COMMAND='sed -e "s/.* t:\([0-9-]\+\).*/\1 &/" -e t -e "s/^/0000-00-00/" | sort -k1 | sed -e "s/^[0-9-]\+ //"'
```

- If you put that into the configuration file, it will apply all the time.
- You can also define a custom add-on (e.g. `listbydue`) if you just need listing of tasks with the changed task order; see [Creating Add ons: Examples](https://github.com/todotxt/todo.txt-cli/wiki/Creating-Add-ons:-Examples) for instructions.
- Alternatively, you could also put that into a custom config file and then define `alias todobydate='todo.sh -d path/to/bydate.cfg'`; that would then apply the changed task order to any subcommand.

It should not be necessary to modify `todo.sh` itself.

**Reply by richmintz1** [reply]
Thank you so much for your clear answer-- I saw "sort" in that line in the config file but I couldn't figure out whether it was the `sort` command or what it was doing.

I already have a layer of add-on commands running (in `.bashrc`), so once I get `sed` doing what I want, I should be able to add one to do this. (There's decades-old regex syntax somewhere deep in my brain.)

**Reply by AvanBerne** [reply]
Thank you guys, this put me on the path to a solution for my use case, where I am using todo.txt as a calendar. For any others searching for similar things, here is my sort command:

```
# Below command sorts by due date, time, and finally priority. Items without due date, time, and/or priority will be sorted last.
export TODOTXT_SORT_COMMAND='sed -e  "s/^.* (\(\w\)) \([0-9:]\+\).* due:\([0-9-]\+\).*/\3 \2 \1 &/" -e t -e "s/^/9999-00-00 00:00 Z /" | sort -k1 | sed -e "s/^[0-9-]\+ [0-9:]\+ [A-Z]\+ //"'
#                                        ^line ^pri    ^time              ^date        ^add date,time,pri,rest   ^add dummy date/time/pri for sorting items that don't match the regex
```

Notes:
* it uses 'due:' instead of 't:' to mark the date.
* it is fragile to cases where the time is not the first thing after the priority, but as I am using a fixed format this works for me.
* I annotated the regex for those who are less familiar with its syntax.
